### DISCUSSION TITLE
No Stacks produced for cc_sampling_rate < 1 Hz

### DISCUSSION THREAD

**Discussion by SValderramaM** [ROBelgium/MSNoise#294] [Q&A] [answered]
Hello everyone,

I am working with MSNoise, it does not show errors when executing each command, but the folders are not created. I have tried changing filters, parameters and could not find the error. The data is in sac format with a duration of 1 day, there are 6 stations with records of 1 year.

**Comment by Frank-kch** [comment]
Hello, 
I encountered same issue when I used monoise stack, do you know how to solve it?
Thank~ 
Frank

**Reply by asyates** [reply]
Hi Frank,

Which version of msnoise are you using? and what does the output text look like when you run the stack command?

Alec

**Reply by Frank-kch** [reply]
Hello Alec,
I'm using version 1.6.2 of MSNoise, , and I can confirm that I am able to successfully run the stack command without any errors, and it does not generate any output text.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/2a44a996-affe-4724-84aa-384d1697e18f)
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/802b8ee1-fddc-456e-b5c5-ea4121335bdb)
Frank

**Reply by asyates** [reply]
mmm the problem isn't immediately apparent to me. 

Any chance you can post screenshots of your config and filter table? and perhaps also a screenshot showing an example of the files in 001_DAYS? Just to check dates match config settings.

This is also with SAC files? I have only ever used MSEED with msnoise, but not aware of any issues with file type.

**Reply by Frank-kch** [reply]
I'm using MSEED files, and I can successfully run compute_cc.

This's a screenshot for the files in 001_DAYS.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/85100410-e033-4f65-8db2-7e899e8f0538)
This' a screenshot for config.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/c5705675-343d-4595-965a-b35c1289ed45)
This' a screenshot for filter table.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/4416fde9-51d5-4881-b36e-3b6e4a99e102)

Thank you very much for your assistance.
Frank

**Reply by asyates** [reply]
Hi Frank,

Thanks for attaching those screenshots; tbh, i'm still not 100% sure what the issue could be based on this. Since it is printing the message 'There are STACKS jobs...', it suggests it is finding jobs with the flag set to 'T' (i assume you are resetting this each time you try?). 

The next message you should get is 'Reading files in...' followed by the filepath, but i guess it's not reaching this step. Basically this part of the code below, where if it reaches get_results() function (from api.py), it should output print("Reading files... in %s" % base):

In s04stack.py:
```
        logger.info(
            "There are STACKS jobs for some days to recompute for %s" % pair)
        sta1, sta2 = pair.split(':')
        for f in filters:
            filterid = int(f.ref)
            for components in params.all_components:
                pair = "%s:%s" % (sta1, sta2)
                sta1 = sta1
                sta2 = sta2
                logger.debug('Processing %s-%s-%i' %
                                  (pair, components, filterid))
                # updated_days = updated_days_for_dates(db, start, end, pair.replace('_', '.'), jobtype='CC', interval=datetime.timedelta(days=interval),returndays=True)
                updated_days = [UTCDateTime(d).datetime.date() for d in days]
                if len(updated_days) != 0:
                    logger.debug("New Data for %s-%s-%i" %
                                  (pair, components, filterid))
                    #~ print updated_days
                    # TODO: load only the updated dates +- max(mov_stack)+1
                    # Note: this would no longer be needed if the stack is h5
                    nstack, stack_total = get_results(
                        db, sta1, sta2, filterid, components, datelist, format=format, params=params)
```
In api.py:
```
def get_results(session, station1, station2, filterid, components, dates,
                mov_stack=1, format="stack", params=None):
                
     // some code

    base = os.path.join("STACKS", "%02i" % filterid,
                        "%03i_DAYS" % mov_stack, components,
                        "%s_%s" % (station1, station2), "%s") + extension
    print("Reading files... in %s" % base)

```
 
Could you maybe run msnoise stack again with the --verbose argument and show output? i.e. after resetting jobs, run 'msnoise -v stack -m'? This should output the info from logger.debug to better see is happening i.e. to see what is stored as pair, components, filterid etc in this loop.

**Reply by Frank-kch** [reply]
Hello Alec,
After I run msnoise -v stack -r, it returns the message "No new data found, hmmm". Could you confirm this problem? According to s04stack.py, it misses nstack. This is a snapshot.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/0402f66a-b5c7-473c-8ee8-ddf7ae4caef1)
I suppose the problem is DA.21413:DA.21414-ZZ-1 because I'm uncertain about the reason why 'ZZ-1' appears after the pair.

However, I've noticed differences between my s04stack.py and your s04stack.py, as well as between my api.py and your api.py. Is this because of different versions?
s04stack.py
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/3df433b8-b7df-4f16-a44a-b7530b9c910b)
api.py
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/02603a12-9e97-48ae-9506-996921aa11ae)

Best regards
Frank

**Reply by asyates** [reply]
Ah woops you're right, i thought i was looking at 1.6.2 codes, but must have swapped back to looking at master version by accident at some point.

**Reply by asyates** [reply]
The 'ZZ-1' looks okay, its from this part of the code:

```
                logger.debug('Processing %s-%s-%i' %
                                  (pair, components, filterid))
```
where you can see it will append components-filterid to the pair for the printed statement, which in this case is ZZ-1.

Seems strange to me that its not finding the data though if compute_cc ran fine i.e. you did not compute CCFs elsewhere (and thus, everything should be in correct format/directory structure). Though... i guess this is the point i would think to check next in case something has gone wrong.

Perhaps adding the line below into the api.py and re-running will help:

```
print("Reading files... in %s" % base)
```


i.e. in api.py, def get_results():
```
    base = os.path.join("STACKS", "%02i" % filterid,
                        "%03i_DAYS" % mov_stack, components,
                        "%s_%s" % (station1, station2), "%s")
    if export_format == "BOTH":
        base += ".MSEED"
        export_format = "MSEED"
    elif export_format == "SAC":
        base += ".SAC"
    elif export_format == "MSEED":
        base += ".MSEED"
print("Reading files... in %s" % base)
```

Then you can check if the file it is trying to read matches the file/directory structure of your STACKS directory.

**Reply by asyates** [reply]
If you do that, you should see something similar to this (form my own codes) printed:

```
Reading files... in STACKS/01/001_DAYS/EN/NZ.WIZ.10_NZ.WIZ.10/%s.MSEED
```

where that final %s placeholder is populated by a date string in the code below:
```
    for j, date in enumerate(dates):
        daystack = base % str(date)
```

e.g.

```
STACKS/01/001_DAYS/EZ/NZ.WIZ.10_NZ.WIZ.10/2021-03-20.MSEED
```

so yeah... i would just double check that the directory structure is all matching correctly. You could even print out the dates string also if you want to double check this is correctly populated too. I have to believe the problem will lie somewhere here because otherwise i'm running out of ideas xD

**Reply by Frank-kch** [reply]
Hello Alec,

 I truly appreciate your assistance. I have followed your instructions to validate the code.
This is the screenshop.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/7d0beeb3-1114-4e5c-a5a5-4b296d575a7d)
The correct directories and files seem to be found.
If you have any solutions in the future, please inform me. I'm still confused as to why the folders are not being created...haha. 
Perhaps I should consider using another program to calculate the dv/v.
Overall, thank you so much.

Sincerely, 
Frank

**Reply by asyates** [reply]
I'm very curious to find the issue now myself so happy to keep digging if you are ;). I have to believe we're getting closer!

The fact the error "No new data found, hmmm" indicates nstack = 0 in the stack04.py codes (below):

stack04.py:
```
                    nstack, stack_total = get_results(
                        db, sta1, sta2, filterid, components, datelist, format=format, params=params)
                    if not nstack:
                        logger.debug("No new data found, hmmm")
```

So in that case, the variable i must not be incrementing in api.py, since it is this variable that is returned as nstack (or just returned as 0 if i==0). So it suggests the read function is failing in the try/except clause (whether due to missing file or another reason).

api.py:
```
    logging.debug("Reading files...")
    for j, date in enumerate(dates):
        daystack = base % str(date)
        try:
            stack_data[j, :] = read(daystack, format=export_format)[0].data[:]
            i += 1
        except:
            pass

    if format == "matrix":
        return i, stack_data

    elif format == "stack":
        logging.debug("Stacking...")

        corr = stack(stack_data, params.stack_method, params.pws_timegate,
                     params.pws_power, params.goal_sampling_rate)

        if i > 0:
            return i, corr
        else:
            return 0, None
 ```
            
Out of interest, have you examined the 1-day CCFs? I noted that your parameter choices are quite different to what i usually am working with (seems you are targeting very long periods?). Wondering if possible there is something wrong with these files and that is why read() is failing (not that i can think of a good reason why that would be right now). 

Otherwise, I might suggest making one further modification to try print out the error message associated with whatever exception is occurring (assuming it must be failing here) to try find out exactly why it is failing. 

So in api.py:

```
for j, date in enumerate(dates):
    daystack = base % str(date)
    try:
        stack_data[j, :] = read(daystack, format=export_format)[0].data[:]
        i += 1
    except Exception as e:
        print(f"An error occurred for date {date}: {e}")
```

**Reply by Frank-kch** [reply]
I believe we have identified the issue.
Our research focuses on Infragravity waves, which exhibit periods closely resembling seismic hum in seismology.
Due to the longer period, we attempt to utilize a 1-day CCF.
We have 5335 data points in the 1-day CCF, in accordance with the sampling rate, and this is a problem.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/0c47fcff-baa9-4943-8bab-c78fff3f4dfd)
Is the limitation of 5334 data points inherent in this program?

Frank

**Reply by ThomasLecocq** [reply]
Hi ! 

No, there is no such limitation, but the main issue I see is that you're using ``corr_duration=86400``, which will, at best, compute 1 window for each day of data. MSNoise is day-based - could you try with 7200s windows or so ?

**Reply by Frank-kch** [reply]
Hello Tomas,

I adjusted corr_duration = 7200s and reran compute_cc and stack, but it doesn't seem to have solved the issue.

Frank

**Reply by asyates** [reply]
Hi Frank,

It is the same error? a difference of 1 in no. data points?

Alec

**Reply by Frank-kch** [reply]
You're correct. I don't think changing corr_duration will affect the data points. It is still 5335.
Is the data point related to the sample rate and analysis_duration?
If I said something wrong please tell me !!

Frank

**Reply by ThomasLecocq** [reply]
am a bit worried about the sampling rate (15 seconds ?) of the input traces and how msnoise treats the 0.06666667 value for resampling, yielding a (random) -1 ; 0 ; +1 samples of the trace...

**Reply by ThomasLecocq** [reply]
also be sure to clean your output folders when recomputing, avoiding stacks of different parametrisations

**Reply by ThomasLecocq** [reply]
sorry, you have to change maxlag of course, to something realistic wrt 7200s data window!

**Reply by Frank-kch** [reply]
Hi Thomas

I tried sample rates of 0.0666666 and 0.06666667. When I used 0.06666667, the STACK folder was not created, indicating that compute_cc did not run successfully. I also adjusted the maxlag to something realistic relative to the 7200s data window (sample rates = 0.0666666, maxlag=3600s). When I ran the stack, the same issue still emerged.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/f4fc0b31-84fa-4d3a-bef6-9ef1c815a0ab)
The data point has indeed changed; you can refer to the screenshot.
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/c8ef8c35-0f1d-4af9-a297-8f0cd6c8766f)
As a result, I think the issue may be related to the sample rate.

Frank

**Reply by ThomasLecocq** [reply]
I see some ways to solve this:
- set cc_sampling_rate to 0.05 with resampling_method=Lanczos in order to avoid a float division
- change the msnoise code so that it would accept cc_sampling_rate = 1/15  and understand that when it contains the "/" symbol, it has to resample to a "delta=15" and not float
- force the array shapes to have the same length (==-> this could be the simplest way, I believe the issue comes from the "t_axis" that contains 1 less sample than the data axis).

**Reply by ThomasLecocq** [reply]
and does the plot ccftime work for the 1 day results ?
Note: your taper length is also too long (that will affect the quality, not the "working" of the code)

**Reply by Frank-kch** [reply]
Hi Thomas

I followed your instructions and successfully resolved the issue. I truly appreciate your assistance. However, I still have a question regarding the compute_dtt. Where should I fill in the interstation?
![image](https://github.com/ROBelgium/MSNoise/assets/122073675/11b63958-ccd3-4e05-bdc6-57542337a165)


Frank

**Reply by ThomasLecocq** [reply]
Yes!! happy!
The interstation is calculated from the station coordinates in the stations table

**Reply by Frank-kch** [reply, marked as answer]
Hi Thomas,

Thank you for your reminder! I will conduct further testing for the taper part later. Regarding the question about 'plot ccftime work,' there is no issue. However, I use this feature less frequently.

Best regards,
Frank
